app/models/fire_cloud_client.rb

```diff
@@ -1332,6 +1336,61 @@ def delete_user_from_billing_project(project_id, role, email)
     end
   end
 
+  ##
+  # PET SERVICE ACCOUNT METHODS
+  # these methods reference the SAM API directly for issuing pet service account tokens/json keyfiles
+  # NOTE: the FireCloudClient instance calling these methods must be initialized using a User object
+  # for authentication purposes, otherwise downstream calls will return 403
+  ##
+
+  # create a new instance of FireCloudClient, but use a pet service account keyfile
+  # will only work with users that have been registered in Terra
+  #
+  # * *params*
+  #   - +user+: (User) => User object from which access tokens are generated
+  #   - +project+: (String) => Default GCP Project to use (can be overridden by other parameters)
+  #
+  # * *return*
+  #   - +FireCloudClient+ instance, or nil if user has not registered with Terra
+  def self.new_with_pet_sa(user, project)
+    # create a temporary client in order to retrieve the user's pet service account keyfile
+    tmp_client = self.new(user, project)
+    if tmp_client.registered?
+      pet_service_account_json = tmp_client.get_pet_service_account_key(project)
+      self.new(user, project, pet_service_account_json)
+    else
+      nil
+    end
+  end
+
+  # issue an access_token for a user's pet service account in the requested project
+  #
+  # * *params*
+  #   - +project_name+ (String) => Name of a FireCloud billing project in which pet service account resides
+  #
+  # * *returns*
+  #   - +String+ pet service account OAuth2 access_token
+  def get_pet_service_account_token(project_name)
+    path = "https://sam.dsde-prod.broadinstitute.org/api/google/v1/user/petServiceAccount/#{project_name}/token"
```

Comment:
I think part of this path should be a constant that lives in a config file.

Reply:
I can do it like `BASE_URL` [here](https://github.com/broadinstitute/single_cell_portal_core/blob/jb-pet-service-accounts/app/models/fire_cloud_client.rb#L18) and call it `SAM_BASE_URL` or `SAM_SERVICE_URL`.  That would at least be internally consistent.

Reply:
f3422a0